rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+        if struct_data is None:
+            struct_data = pkg_resources.resource_string("rig",
+                                                        "boot/sark.struct")
+        self.struct_data = struct_data
+
+        if boot_data is None:
+            boot_data = pkg_resources.resource_string("rig", "boot/scamp.boot")
+        self.boot_data = boot_data
+
+        # Create the initial connection
+        self.connections = [
+            SCPConnection(initial_host, n_tries, timeout)
+        ]
+
+    def __call__(self, **context_args):
+        """Create a new context for use with `with`.
+
+        E.g:
+            with controller(x=3, y=4):
+                # All commands will now communicate with chip (3, 4)
+        """
+        return self.get_new_context(**context_args)
+
+    def send_scp(self, x, y, *args, **kwargs):
+        """Determine the best connection to use to send an SCP packet and use
+        it to transmit.
+
+        See the arguments for
+        :py:method:`~rig.machine_controller.scp_connection.SCPConnection` for
+        details.
+        """
+        return self.connections[0].send_scp(x, y, *args, **kwargs)
+
+    def boot(self, width, height, **boot_kwargs):
+        """Boot a SpiNNaker machine of the given size.
+
+        Parameters
+        ----------
+        width : int
+            Width of the machine (0 < w < 256)
+        height : int
+            Height of the machine (0 < h < 256)
+
+        For further boot arguments see
+        :py:func:`~rig.machine_controller.boot.boot`.  **Regardless** of the
+        values passed to this method the struct and boot data from `__init__`
+        will be used.
+
+        Notes
+        -----
+        The constants `rig.boot.spinX_boot_options` can be used to specify boot
+        parameters, for example:
+
+            controller.boot(**spin5_boot_options)
+
+        Will boot the Spin5 board connected to by `controller`.
+        """
+        boot_kwargs.update({
+            "boot_data": self.boot_data,
+            "struct_data": self.struct_data,
+        })
+        boot.boot(self.initial_host, width=width, height=height, **boot_kwargs)
+
+    @ContextMixin.use_contextual_arguments
+    def software_version(self, x=Required, y=Required, processor=0):
+        """Get the software version for a given SpiNNaker core.
+        """
+        sver = self.send_scp(x, y, processor, SCPCommands.sver)
+
+        # Format the result
+        # arg1 => p2p address, physical cpu, virtual cpu
+        p2p = sver.arg1 >> 16
+        pcpu = (sver.arg1 >> 8) & 0xff
+        vcpu = sver.arg1 & 0xff
+
+        # arg2 => version number and buffer size
+        version = (sver.arg2 >> 16) / 100.
+        buffer_size = (sver.arg2 & 0xffff)
+
+        return CoreInfo(p2p, pcpu, vcpu, version, buffer_size, sver.arg3,
+                        sver.data)
+
+    @ContextMixin.use_contextual_arguments
+    def write(self, address, data, x=Required, y=Required, p=Required):
```

Comment:
Is p=0 not a sensible default here? (e.g. when writing to SDRAM, the common case I would guess, there is no reason to do so via an application core?)